### DISCUSSION TITLE
Actix vs uWebSockets

### DISCUSSION THREAD

**Discussion by yamiteru** [uNetworking/uWebSockets.js#445] [General]
What is faster?

**Comment by ghost** [comment]
If you look under benchmarks folder in main repo you can see where actix-web falls. But I would probably guess anything Node.js to fall short. These results are given the C++ library without anything Node.js involved.

**Reply by ghost** [reply]
Honestly, if you **do have the ability** to work with Rust instead of Node.js I would pick Rust any day of the week.

**Comment by yamiteru** [comment]
Hmm I see. Actix-web seems to be less than 50% performant when compared to uWebSockets. Is it possible to use C++ uWebSockets in Rust?

**Reply by yamiteru** [reply]
Also the benchmarks in the main repo are 2 years old.

**Reply by ghost** [reply]
If the ones provided are not to your liking you can do your own. I don't see Actix providing benchmarks of us, so we are already over-extending.

**Reply by yamiteru** [reply]
Is it possible to use C++ uWebSockets in Rust?

**Reply by ghost** [reply]
No, Swift, Rust and C++ are not immediately compatible but they are all compatible with C, which is the only way to bridge between systems programming languages in a compatible way. So you would have to write a C-wrapper and then a Rust wrapper of that C wrapper.

**Reply by yamiteru** [reply]
Ok, thank you very much. You're very kind. So I'm gonna switch to Rust and Actix-web since I've wanted to learn Rust for a long time. But don't worry. I'm using uWebSockets in all the other projects 😉 So I want to thank you for giving us this really cool library and for being you. I like your point on middlewares. They're just functions that run in the handler, so why add special functions for it? 👌

**Reply by ghost** [reply]
I think you're going to have no issues with Actix web.

uWebSockets mostly has value for people stuck with nodejs, or with pedantic latency needs, etc.

The returns are definitely diminishing at higher perf.

**Reply by hst-m** [reply]
I looked into Rust couple months ago, confirmed there are large benefits of simple language like JS, allows you to focus on things other than details of language, I need simple with least amount of code so I can do app logic on front and back end, so uWS.js is great for that

**Reply by ghost** [reply]
That's personal taste. I for sure am more productive in C++ than in JavaScript, esp when you have larger projects and need to debug/refactor. Debugging/refactoring large JS projects is self hatred.

**Reply by yamiteru** [reply]
Even middle sized projects in JS are hell. I use TS to counter it. But still it compiles to a hellish unsafe language. I really like statically typed languages. They are harder to learn for sure, but they give a lot of control in return. For me it's crucial to have a lot of control and abstract only where I need to.

BTW just out of curiosity - Could theoretically be uWebSockets be completely rewritten in Rust or is there something that would prevent it?

**Reply by ghost** [reply]
I agree but the existence and popularity of this project kind of validates the fact that some people really do like JavaScript and rather "hack around" its shortcomings than just pick something more suited for the task. To me, when someone **likes** JavaScript that is kind of like looking at this documentary: https://www.youtube.com/watch?v=x5OiplprDFU

A girl that's in sexual relations with a steel bridge. I say go for it if you like it, but I won't follow. I accept your oddities, but yuck! :D

**Reply by hst-m** [reply]
I think you all are forgetting, the #1 platform for browsing the web is something called a Web Browser that uses Javascript as the coding language. For those of us who do web browser coding including the majority of all apps, it is helpful to code the server in the same language as the browser so we can focus on doing simple things right in the same language. I would totally learn C/C+++ and Rust if I had extra time but it may not happen currently. I am a solo developer who does the full stack so can't export this out to someone else in different language

**Reply by ghost** [reply]
It's okay if you're into bridges.

**Reply by ghost** [reply]
https://youtu.be/x5OiplprDFU?t=1357

**Reply by hst-m** [reply]
I do enjoy coding with JS so it is possible there is something wrong with me. I like coding the least amount of code possible and then compress it down like paragraphs in a book. Good thing I don't have to work with other people they might not like my style, but I probably use x15 less code than the monstrous React apps they were producing at previous company that were filled with bugs

**Reply by yamiteru** [reply]
Hey so I'm back after 4 hours of trying to set up MongoDB in Rust and idk if I'm dumb, but hell Rust is super strict. The async/await is really weird. Wtf is unwrap and why do I have to use it on like every second line? I just can't make the MongoDB work. But setting up Actix was done in like 10 minutes. Also Rust plugin for IntelliJ sucks. Even VS Code is better and that's a freaking text editor and not IDE.

BTW I'm a fullstack JS/TS dev too. I usually use StencilJS/Svelte on frontend and uWebSockets on backend.

I like the fact that JS is easy to start with. I wish it had types tho. That's why I use TS.

**Reply by hst-m** [reply]
I am big Mongo fan use it in all my apps. No Typescript though because it adds extra code and clutters it up. Sublime Text 3 for all text editing and developing needs and Sublime Merge is awesome for Git UI

**Reply by yamiteru** [reply]
Yeah I use MongoDB quite a lot too. It's easy to understand and gets the job done. For big data apps and website I use JanusGraph with ScyllaDB as a storage.

Have you tried TS? I think the pros are bigger than cons. I like a lean code too. If you separate types, enums, abstract classes, etc. into their own files then there is not much clutter around.

I write my code in WebStorm and I use Fork as a Git GUI client.
